google/cloud/forseti/scanner/audit/kms_rules_engine.py

```diff
@@ -346,34 +338,107 @@ def __init__(self, rule_name, rule_index, rule):
+            bool: Returns true if a match is found.
         """
-        crypto_key = self.rule['key']
-        for key_data in crypto_key:
-            key_rotation_period = key_data.get('rotation_period')
         LOGGER.debug('Formatting rotation time...')
+        creation_time = key.primary_version.get('createTime')
+        scan_time = date_time.get_utc_now_datetime()
         last_rotation_time = creation_time[:-5]
         formatted_last_rotation_time = datetime.datetime.strptime(
             last_rotation_time, string_formats.TIMESTAMP_MICROS)
         days_since_rotated = (scan_time - formatted_last_rotation_time).days
+        print('days:', days_since_rotated)
+        if mode == BLACKLIST and days_since_rotated > rotation_period:
+            return True
+        elif mode == WHITELIST and days_since_rotated <= rotation_period:
+            return True
+        return False
+
+    @classmethod
+    def find_match_algorithms(cls, key, rule_algorithms):
+        """Check if there is a match for this rule algorithm against the given
+        resource.
+
+        Args:
+            key (Resource): The resource to check for a match.
+            rule_algorithms (string): The algorithms of this rule.
 
-        if days_since_rotated > key_rotation_period:
-            return False
+        Returns:
+            bool: Returns true if a match is found.
+        """
+        LOGGER.debug('Finding rule algorithm in key algorithm')
+        key_algorithm = key.primary_version.get('algorithm')
+        for algorithm in rule_algorithms:
+            if key_algorithm == algorithm:
+                return True
+        return False
+
+    def find_match_protection_level(self, key, rule_protection_level):
+        """Check if there is a match for this rule protection level against the
+         given resource.
+
+        Args:
+            key (Resource): The resource to check for a match.
+            rule_protection_level (string): The protection level of this rule.
+
+        Returns:
+            bool: Returns true if a match is found.
+        """
+        key_protection_level = key.primary_version.get('protectionLevel')
+        if key_protection_level == rule_protection_level:
+            return True
+        return False
 
-        return True
+    def find_match_purpose(self, key, rule_purpose):
+        """Check if there is a match for this rule purpose against the given
+        resource.
+
+        Args:
+            key (Resource): The resource to check for a match.
+            rule_purpose (string): The purpose of this rule.
+
+        Returns:
+            bool: Returns true if a match is found.
+        """
+        key_purpose = key.purpose
+        symmetric_algorithm = 'ENCRYPT_DECRYPT'
+        if rule_purpose == 'symmetric':
+            if key_purpose == symmetric_algorithm:
+                return True
+        elif not key_purpose == symmetric_algorithm:
```

Comment:
I dont understand this part, why arent you matching key_purpose with rule_purpose, instead you are putting in custom strings like 'ENCRYPT_DECRYPT' or 'symmetric' in here

Reply:
The use case for this method is to check if a key is symmetric or asymmetric. The valid values of `purpose` in rule is `symmetric` and `asymmetric`. The only purpose allowed for symmetric key encryption is `ENCRYPT_DECRYPT`. So I'm checking against that.
https://cloud.google.com/kms/docs/algorithms